rfcs/0024-reducing-time-per-transaction.md

```diff
@@ -0,0 +1,77 @@
+
+```
+verification time per SNARK * number of SNARKs per transaction
+```
+
+Currently the number of SNARKs per transaction is at least 2 and the verification time is about 155ms (on my laptop). This means that at 1 TPS, SNARK verification eats up at least 310ms of our maximum TPT of 1 second.
+
+Luckily both these numbers can be substantially pushed down. 
+
+### Reducing number of SNARKs per transaction
+
+The number of SNARKs per transaction is equal to the number of nodes in a parallel scan state tree divided by the number of transactions in that tree.
+
+Let us say the number of transactions is N. Say there are B transactions per base proof and that the branching factor of the merge nodes is K.
+
+Then the total number of nodes in a scan state tree is
+
+```
+(N / B) + \sum_{i=0}^{log_K(N / B) - 1} K^i
+```
+
+It's not totally obvious but the sum is a value which gets smaller with larger K. Thus there are two ways we can reduce the number of SNARKs per transaction, and in so doing reduce our overall TPS.
+
+1. Increase the bundle size B of number of transactions per base proof. Right now it is 1. I think it would be reasonable to make it 4.
+2. Increase the branching factor of the scan state.
+
+### Reducing verification time per SNARK
+
+I believe we should not concentrate on this right now because the underlying SNARK may change substantially in the next month. That said, I will record some of what I know here.
+
+Right now verification time for the SNARK is about 155ms (on my laptop) and looks like
+
+```
+hashing: 3ms
+g2 subgroup check: 26.5ms * 2
+preprocess vk: 29ms
+G2 precomp: 14ms * 2
+miller loop: 8ms
+double miller loop: 14ms * 2
+final exp: 2ms * 2
+```
+
+Most of these checks can be batched across a list of proofs.
+
+- The "preprocess vk" part can be done once at application launch and so can be eliminated entirely.
+- The "g2 subgroup check" can be batched across proofs such that the marginal time is brought down to about 4.5ms per proof.
+- Of the five miller loops, 2 can be eliminated. This means about 24ms for miller loops marginally.
+- The final exp can be eliminated.
+
+So at the end, the marginal time can probably be brought down to about (4.5ms + 24ms + 3ms + 28ms) = 59.5ms.
+
+## Priorities and recommendations
+
+The biggest, easiest win would be to increase the bundle size B, the number of transactions per base proofs.
+If we increase it to 4, we reduce the SNARK-verification part of our TPT by 1/4. 
+
+I think it is not worth spending time implementing batch verification right now because our SNARK construction is likely to change substantially soon.
+
+Increasing the branching factor is worth doing, but it is worth noting it at most improves SNARK-TPT  by a factor of 2. (You can see this by the equation for the number of nodes in a scan state tree. The sum is (N/B - 1) when branching factor is 2, as it is now, which means we still have N/B even if that sum gets eliminated entirely.
```

Comment:
A point to note when deciding the value of `B` and `K` is that Increasing `B` and `K` impacts proof generation time and that in turn affects the feasibility of proof generation. Currently generating a proof bundle takes time `t` <= block window duration and if we need `X` proof bundles per block we need at least `X` snark workers.  Increasing `B` and `K` would increase the base proof and merge proof generation time respectively. Increase in `t => Increase in number of snark workers and Increase in delay` so that there's enough work accumulated and there are enough snark workers to do them and therefore have required number of proofs per block.